src/search/tasks/negated_axioms_task.cc

```diff
@@ -0,0 +1,323 @@
+                needed_negatively.insert(var2);
+                to_process.push_back(var2);
+            }
+        }
+    }
+
+    // If there are no derived variables that occur as their default value we don't need to do anything.
+    if (needed_negatively.empty()) {
+        return;
+    }
+
+    // Build dependency graph and collect relevant axioms for derived values occurring as their default value.
+    // TODO: The dependency information is a superset of positive_dependencies from above.
+    // TODO: The graph includes also (unreachable) vertices for non-derived variables.
+    vector<vector<int>> dependency_graph;
+    dependency_graph.resize(task_proxy.get_variables().size(), {});
+
+    for (OperatorProxy axiom: task_proxy.get_axioms()) {
+        FactPair head = axiom.get_effects()[0].get_fact().get_pair();
+        for (FactProxy condition: axiom.get_effects()[0].get_conditions()) {
+            if (condition.get_variable().is_derived()) {
+                dependency_graph[condition.get_pair().var].push_back(head.var);
+            }
+        }
+    }
+
+    // Get strongly connected components and map all relevant vars to their scc
+    vector<vector<int>> sccs = sccs::compute_maximal_sccs(dependency_graph);
+    unordered_map<int, int> var_to_scc;
+    for (size_t i = 0; i < sccs.size(); ++i) {
+        for (int var: sccs[i]) {
+            if (needed_negatively.count(var) > 0) {
+                var_to_scc.insert({var, i});
+            }
+        }
+    }
+
+    for (int var: needed_negatively) {
+        vector<OperatorProxy> &axioms = axioms_for_var[var];
+
+        if (sccs[var_to_scc[var]].size() > 1) {
+            /* If the cluster contains multiple variables, they have a cyclic
+               positive dependency. In this case, the "obvious" way of
+               negating the formula defining the derived variable is
+               semantically wrong. For details, see issue453.
+
+               Therefore, in this case we perform a naive overapproximation
+               instead, which assumes that derived variables occurring in
+               such clusters can be false unconditionally. This is good
+               enough for correctness of the code that uses these negated
+               axioms, but loses accuracy. Negating the rules in an exact
+               (non-overapproximating) way is possible but more expensive.
+               Again, see issue453 for details.
+            */
+            std::string name = task_proxy.get_variables()[var].get_name() + "_negated";
+            // TODO: fix the hardcoded 1
+            negated_axioms.emplace_back(FactPair(var, 1), vector<FactPair>(), name);
+        } else {
+            // TODO: fix the hardcoded 1
+            add_negated_axioms(FactPair(var, 1), axioms, task_proxy);
+        }
+    }
+    cout << negated_axioms.size() << " negated axioms" << endl;
+}
+
+void NegatedAxiomsTask::find_non_dominated_hitting_sets(
```

Comment:
This is a recursive function to multiply out `conditions_as_cnf`. It could also be implemented as a while loop instead, but I found this more readable.
(I would probably want to change the name though, but don't have a good idea at the moment.)